purge/purge.py

```diff
@@ -0,0 +1,330 @@
+            admin_password = str(pk.sign_msg_hash(eth_sign_hash(server_name.encode())))
+            if admin_private_key_print_only:
+                click.secho(f'PK to Matrix User:     {admin_user}')
+                click.secho(f'PK to Matrix Password: {admin_password}')
+                return
+
+        if admin_user:
+            if not admin_password:
+                admin_password = click.prompt(
+                    'Admin user password',
+                    default='',
+                    hide_input=True,
+                )
+            response = session.post(
+                urljoin(server, '/_matrix/client/r0/login'),
+                json={
+                    'type': 'm.login.password',
+                    'user': admin_user,
+                    'password': admin_password,
+                },
+            )
+            assert response.status_code == 200, f'{response!r} => {response.text!r}'
+            admin_access_token = response.json()['access_token']
+            with open(admin_access_token_file, 'w') as fo:
+                json.dump({
+                    'admin_user': admin_user,
+                    'admin_access_token': admin_access_token,
+                }, fo, indent=2)
+
+        else:
+            raise RuntimeError('No admin_user nor previous access token found')
+
+    with psycopg2.connect(db_uri) as db, db.cursor() as cur:
+
+        # set user as admin in database if needed
+        cur.execute(
+            'SELECT admin FROM users WHERE name = %s ;',
+            (admin_user,),
+        )
+        if not cur.rowcount:
+            raise RuntimeError(f'User {admin_user!r} not found')
+        is_admin, = cur.fetchone()
+        if admin_set and not is_admin:
+            cur.execute(
+                'UPDATE users SET admin=1 WHERE name = %s ;',
+                (admin_user,),
+            )
+            db.commit()
+        elif not is_admin:
+            raise RuntimeError(f'User {admin_user!r} is not an admin. See --admin-set option')
+
+        purges = dict()
+        def wait_and_purge_room(room_id=None, event_id=None):
+            """ Wait for available slots in parallel_purges and purge room
+
+            If room_id is None, just wait for current purges to complete and return
+            If event_id is None, purge all events in room
+            """
+            while len(purges) >= (parallel_purges if room_id else 1):
+                # wait and clear completed purges
+                time.sleep(1)
+                for _room_id, purge_id in list(purges.items()):
+                    response = session.get(
+                        urljoin(
+                            server,
+                            '/_matrix/client/r0/admin/purge_history_status/' + quote(purge_id),
+                        ),
+                        params={'access_token': admin_access_token},
+                    )
+                    if response.status_code != 200 or response.json().get('status') != 'active':
```

Comment:
Can this fail? If it can then we should at least log errors here.

Reply:
I don't think so, just catched it here to clean if it happens, to avoid a crash, but will add a log as well